ddl/index.go

```diff
@@ -815,59 +808,45 @@ func (r *reorgIndexTask) String() string {
 	return "physicalTableID" + strconv.FormatInt(r.physicalTableID, 10) + "_" + "[" + r.startHandle.String() + "," + r.endHandle.String() + rightParenthesis
 }
 
-type addIndexResult struct {
+type backfillResult struct {
 	addedCount int
 	scanCount  int
 	nextHandle kv.Handle
 	err        error
 }
 
-// addIndexTaskContext is the context of the batch adding indices.
-// After finishing the batch adding indices, result in addIndexTaskContext will be merged into addIndexResult.
-type addIndexTaskContext struct {
+// backfillTaskContext is the context of the batch adding indices.
+// After finishing the batch adding indices, result in backfillTaskContext will be merged into backfillResult.
+type backfillTaskContext struct {
 	nextHandle kv.Handle
 	done       bool
 	addedCount int
 	scanCount  int
 }
 
 // mergeAddIndexCtxToResult merge partial result in taskCtx into result.
-func mergeAddIndexCtxToResult(taskCtx *addIndexTaskContext, result *addIndexResult) {
+func mergeAddIndexCtxToResult(taskCtx *backfillTaskContext, result *backfillResult) {
 	result.nextHandle = taskCtx.nextHandle
 	result.addedCount += taskCtx.addedCount
 	result.scanCount += taskCtx.scanCount
 }
 
 func newAddIndexWorker(sessCtx sessionctx.Context, worker *worker, id int, t table.PhysicalTable, indexInfo *model.IndexInfo, decodeColMap map[int64]decoder.Column) *addIndexWorker {
 	index := tables.NewIndex(t.GetPhysicalID(), t.Meta(), indexInfo)
-	rowDecoder := decoder.NewRowDecoder(t, decodeColMap)
+	rowDecoder := decoder.NewRowDecoder(t, t.WritableCols(), decodeColMap)
 	return &addIndexWorker{
-		id:          id,
-		ddlWorker:   worker,
-		batchCnt:    int(variable.GetDDLReorgBatchSize()),
-		sessCtx:     sessCtx,
-		taskCh:      make(chan *reorgIndexTask, 1),
-		resultCh:    make(chan *addIndexResult, 1),
-		index:       index,
-		table:       t,
-		rowDecoder:  rowDecoder,
-		priority:    kv.PriorityLow,
-		defaultVals: make([]types.Datum, len(t.Cols())),
-		rowMap:      make(map[int64]types.Datum, len(decodeColMap)),
-	}
-}
-
-func (w *addIndexWorker) close() {
-	if !w.closed {
-		w.closed = true
-		close(w.taskCh)
+		backfillWorker: newBackfillWorker(sessCtx, worker, id, t),
+		index:          index,
+		rowDecoder:     rowDecoder,
+		defaultVals:    make([]types.Datum, len(t.WritableCols())),
+		rowMap:         make(map[int64]types.Datum, len(decodeColMap)),
 	}
 }
 
 // getIndexRecord gets index columns values from raw binary value row.
 func (w *addIndexWorker) getIndexRecord(handle kv.Handle, recordKey []byte, rawRecord []byte) (*indexRecord, error) {
 	t := w.table
-	cols := t.Cols()
+	cols := t.WritableCols()
```

Comment:
Does this change affect the correctness of the usage of `cols`?

For example, there is an offset related usage at line 858 below.
```go
col := cols[v.Offset]
```

Reply:
Yes.